File: Typecast/TCCmapFormat0.m
//
//  TCCmapFormat0.m
//  Typecast
//
//  Created by David Schweinsberg on 3/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCCmapFormat0.h"
#import "TCDataInput.h"

@interface TCCmapFormat0 ()

@property (strong) NSArray *glyphIdArray;

@end


@implementation TCCmapFormat0

- (id)initWithDataInput:(TCDataInput *)dataInput
{
    self = [super initWithDataInput:dataInput];
    if (self)
    {
        [self setRanges:@[[NSValue valueWithRange:NSMakeRange(0, 255)]]];

        [self setFormat:0];
        NSMutableArray *glyphIdArray = [[NSMutableArray alloc] initWithCapacity:256];
        for (int i = 0; i < 256; ++i)
            [glyphIdArray addObject:[NSNumber numberWithUnsignedChar:[dataInput readUnsignedByte]]];
        _glyphIdArray = glyphIdArray;
    }
    return self;
}

- (NSUInteger)glyphCodeAtCharacterCode:(NSUInteger)characterCode
{
    if (characterCode < 256)
        return [_glyphIdArray[characterCode] unsignedIntegerValue];
    else
        return 0;
}

@end

File: Typecast/TCCmapFormat0.swift
//
//  TCCmapFormat0.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/6/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation

/**
 * Simple Macintosh cmap table, mapping only the ASCII character set to glyphs.
 */
class TCCmapFormat0: TCCmapFormat {
  let length: Int
  let language: Int
  let glyphIdArray: [Int]

  init(dataInput: TCDataInput) {
    length = Int(dataInput.readUInt16())
    language = Int(dataInput.readUInt16())
    var glyphIdArray = [Int]()
    for _ in 0...255 {
      glyphIdArray.append(Int(dataInput.readUInt8()))
    }
    self.glyphIdArray = glyphIdArray
  }

  var format: Int {
    get {
      return 0
    }
  }

  var ranges: [CountableClosedRange<Int>] {
    return [0...255]
  }

  func glyphCode(characterCode: Int) -> Int {
    if characterCode < 256 {
      return glyphIdArray[characterCode]
    } else {
      return 0
    }
  }
}
